source/extensions/filters/network/thrift_proxy/metadata.h

```diff
@@ -134,8 +136,10 @@ class MessageMetadata {
   /**
    * @return HeaderMap of current headers (never throws)
    */
-  const Http::HeaderMap& headers() const { return *headers_; }
-  Http::HeaderMap& headers() { return *headers_; }
+  const Http::RequestHeaderMap& requestHeaders() const { return *request_headers_; }
+  Http::RequestHeaderMap& requestHeaders() { return *request_headers_; }
+  const Http::ResponseHeaderMap& responseHeaders() const { return *response_headers_; }
+  Http::ResponseHeaderMap& responseHeaders() { return *response_headers_; }
```

Comment:
Does a given Metadata instance ever have both request and response headers? Maybe we can do away with the distinction here? I think in key places we might want to assert that the metadata is from a client request vs an upstream response.

Reply:
I don't think a metadata can ever hold both. The reason I went with this approach is to avoid games around the type system (RequestHeaderMap vs ResponseHeaderMap). We can't use plain `HeaderMap`, because `access_log->log()` explicitly requires `RequestHeaderMap` and `ResponseHeaderMap` (we don't care about trailers). And I rather avoid `absl::any` or similar tricks. 

What were you thinking?

Reply:
Maybe https://github.com/envoyproxy/envoy/pull/21276/commits/865e9d029a0043d2aed2d7bf91de030606fb7cc8 is enough?

Reply:
Wonderful, this change caught some potential bugs: the encoder filter expects the metadata to hold both request and response headers, which shouldn't be allowed (I think). I'll look in a bit.